modules/tests/shared/src/test/scala/DescribeCacheTest.scala

```diff
@@ -82,6 +83,67 @@ class DescribeCacheTest extends SkunkTest {
     } yield "ok"
   }
 
+  val runs = 100
+  // This should not fail 
+  pooledTest("portal1 - concurrent portal with normal flatMap") { pool =>
+    import scala.concurrent.duration._
+    def cmd(idx: Int): Command[String *: Int *: EmptyTuple] = sql"INSERT INTO scalars VALUES (${text.varchar}, ${int4}) -- #${idx.toString}".command
+    1.to(runs).toList.parTraverse{ idx =>
+      pool.use { s =>
+        s.transaction.use { _ =>
+          s.prepare(cmd(idx)).flatMap(_.execute((s"name$idx", idx))).flatMap (_ => 
+            IO.sleep(1.second) >> s.prepare(cmd(idx)).flatMap(_.execute((s"name$idx", idx))).void
+          )
+        }
+      }
+    } 
+  }
+  
+  // This should not fail 
+  pooledTest("portal2 - concurrent portal with map / identity *inside tx*") { pool =>
+    import scala.concurrent.duration._
+    def cmd(idx: Int): Command[String *: Int *: EmptyTuple] = sql"INSERT INTO scalars VALUES (${text.varchar}, ${int4}) -- #${idx.toString}".command
+    1.to(runs).toList.parTraverse{ idx =>
+      pool.use { s =>
+        s.transaction.use { _ =>
+          s.prepare(cmd(idx)).flatMap(_.execute((s"name$idx", idx))).map (_ => 
+            IO.sleep(1.second) >> s.prepare(cmd(idx)).flatMap(_.execute((s"name$idx", idx))).void
+          ).flatMap(identity)
+        }
+      }
+    } 
+  }
+
+  // This will fail if run enough 
+  pooledTest("portal3 - concurrent portal with map / identity *outside tx*") { pool =>
+    import scala.concurrent.duration._
+    def cmd(idx: Int): Command[String *: Int *: EmptyTuple] = sql"INSERT INTO scalars VALUES (${text.varchar}, ${int4}) -- #${idx.toString}".command
+    1.to(runs).toList.parTraverse{ idx =>
+      pool.use { s =>
+        s.transaction.use { _ =>
+          s.prepare(cmd(idx)).flatMap(_.execute((s"name$idx", idx))).map (_ => 
+            IO.sleep(1.second) >> s.prepare(cmd(idx)).flatMap(_.execute((s"name$idx", idx))).void
+          )
+        }
+      }.flatMap(identity)
+    }
+  }
+  
+  // This will fail if run enough 
+  pooledTest("portal4 - concurrent portal with flatten *outside tx*") { pool =>
+    import scala.concurrent.duration._
+    def cmd(idx: Int): Command[String *: Int *: EmptyTuple] = sql"INSERT INTO scalars VALUES (${text.varchar}, ${int4}) -- #${idx.toString}".command
+    1.to(runs).toList.parTraverse{ idx =>
+      pool.use { s =>
+        s.transaction.use { _ =>
+          s.prepare(cmd(idx)).flatMap(_.execute((s"name$idx", idx))).map (_ => 
+            IO.sleep(1.second) >> s.prepare(cmd(idx)).flatMap(_.execute((s"name$idx", idx))).void
+          )
+        }
+      }.flatten
```

Comment:
`.use(...).flatten` looks wrong. It means you are building an `IO` in the `use(...)` when the `Resource` is available, but you are executing it _after_ the `use`, when the resource has been released.